source/common/common/packed_struct.h

```diff
@@ -0,0 +1,137 @@
+ *      auto& element0 = packed_struct.get<MyElements::element0>();
+ *    }
+ * ```
+ */
+template <class T, uint8_t max_size, class ElementName> class PackedStruct {
+public:
+  PackedStruct(size_t a_capacity = 0)
+      : data_(a_capacity > 0 ? std::make_unique<T[]>(a_capacity) : nullptr) {
+    static_assert(std::is_enum_v<ElementName>);
+    static_assert(max_size > 0);
+
+    RELEASE_ASSERT(a_capacity < std::numeric_limits<uint8_t>::max(),
+                   "capacity should fit in uint8_t.");
+    indices_.fill(max_size);
+    indices_[max_size] = a_capacity;
+  }
+
+  // Accessors.
+  template <ElementName element_name> T& get() {
+    sizeCheck<element_name>();
+    hasCheck<element_name>();
+    return (data_.get())[indices_[static_cast<std::underlying_type_t<ElementName>>(element_name)]];
+  }
+  template <ElementName element_name> const T& get() const {
+    return const_cast<const T&>(get<element_name>());
+  }
+
+  // Check to see if element is populated.
+  template <ElementName element_name> bool has() const {
+    sizeCheck<element_name>();
+    return indices_[static_cast<std::underlying_type_t<ElementName>>(element_name)] < max_size;
+  }
+
+  // Set element.
+  template <ElementName element_name> void set(T t) {
+    sizeCheck<element_name>();
+
+    static constexpr auto element_idx =
+        static_cast<std::underlying_type_t<ElementName>>(element_name);
+    auto const current_size = size();
+    // If we're at capacity and we don't have a slot for element_idx, increase capacity by 1.
+    if (!has<element_name>() && current_size == capacity()) {
+      auto tmp = std::make_unique<T[]>(++indices_[max_size]);
+      std::move(data_.get(), std::next(data_.get(), current_size), tmp.get());
+      data_ = std::move(tmp);
+    }
+    if (!has<element_name>()) {
+      indices_[element_idx] = current_size;
+    }
+    data_.get()[indices_[element_idx]] = t;
+  }
+
+  // Number of non-empty elements in the struct. Note that this can be less than
+  // capacity.
+  size_t size() const {
+    return std::count_if(indices_.begin(), indices_.end() - 1,
+                         [](uint8_t a) { return a < max_size; });
+  }
+
+  // Disable copying
+  PackedStruct(const PackedStruct&) = delete;
+
+  // Move constructor and assignment operator.
+  PackedStruct(PackedStruct&& other) noexcept = default;
+  PackedStruct& operator=(PackedStruct&& other) = default;
+
+  friend class PackedStructTest;
+
+private:
+  size_t capacity() const { return indices_[max_size]; }
+
+  template <ElementName element_name> static constexpr void sizeCheck() {
+    static_assert(static_cast<size_t>(element_name) < static_cast<size_t>(max_size));
+  }
+  template <ElementName element_name> void hasCheck() const {
+    ENVOY_BUG(has<element_name>(),
+              absl::StrCat("Element corresponding to index ",
+                           static_cast<std::underlying_type_t<ElementName>>(element_name),
+                           " is not assigned"));
```

Comment:
If this fails, the calling code is almost certainly going to enter undefined behavior and do something bad. Shouldn't this outright crash or if that is not desirable instead of has() just have get() return a StatusOr and have the caller decide what to do?

Reply:
I had a `RELEASE_ASSERT` before :-). I see how `StatusOr` makes this much more cleaner. Will implement that.

Reply:
I used OptRef instead. PTAL.